rust_src/src/keyboard.rs

```diff
@@ -32,6 +34,66 @@ use crate::{
 #[cfg(feature = "window-system")]
 use crate::remacs_sys::cancel_hourglass;
 
+#[derive(Clone, Copy, Debug)]
+pub struct Event(LispObject);
+
+impl Event {
+    // replaces EVENT_HAS_PARAMETERS
+    /// Checks if the event has data fields describing it (i.e. a mouse click).
+    pub fn has_parameters(self) -> bool {
+        self.0.is_cons()
+    }
+
+    /// Checks if the event's data fields have contents
+    pub fn has_data(self) -> bool {
+        match self.0.as_cons() {
+            Some(cons) => cons.cdr().is_cons(),
+            None => false,
+        }
+    }
+
+    // replaces EVENT_HEAD
+    /// Extract the head from an event.
+    /// This works on composite and simple events.
+    pub fn head(self) -> LispObject {
+        match self.0.as_cons() {
+            Some(cons) => cons.car(),
+            None => self.0,
+        }
+    }
+
+    // replaces EVENT_HEAD_KIND
+    // Note that the C macro expects an event head, this method expects an event which it exctracts the head from itself.
+    /// Extracts the kind from an event's head.
+    pub fn head_kind(self) -> LispObject {
+        lists::get(self.head().into(), Qevent_kind)
+    }
+
+    // replaces EVENT_START
+    /// Extracts the starting position from a composite event.
+    pub fn start(self) -> LispObject {
+        car_safe(cdr_safe(self.0))
```

Comment:
These could be a little clearer with pattern matching. Something like

```rust
    match self.0.into() {
        Some((_, (appropriate_name, _)) => appropriate_name,
        None => Qnil,
    }
```

Is that right? Probably not, but you get the idea. That way the structure of the object is easier to see.

Similar for all other uses of car / cdr. Really, there are few reasons for them to be used in Rust at all.

Reply:
The issue with pattern matching is we need a car and then a cdr and either may not be there. So the code above would need to be something like:
```
match self.0.into() {
        Some(((_, appropriate_name), _)) => appropriate_name,
        None => Qnil,
}
```

I am not sure that works. Since it needs to evaluate to None when any of the internal transformations do not work.

Reply:
I'm starting to doubt this works for method works for substituting several car/cdr calls at all unless we add some more From impl's, and even then it might not.

Calling `.into()` once makes the compiler complain that the inner tuple is not a `LispObject`, and calling it twice makes the type of the first into call uninferrable, and adding  any type annotations does not fix it.

Reply:
I get it mixed up sometimes -- we can convert arbitrarily nested tuples in LispObjects, but not vice versa. It would be nice though.